src/ansi-c/goto-conversion/goto_clean_expr.cpp

```diff
@@ -440,9 +476,155 @@ goto_convertt::clean_expr_resultt goto_convertt::clean_expr(
   }
   else if(expr.id() == ID_forall || expr.id() == ID_exists)
   {
+    quantifier_exprt &qex = to_quantifier_expr(expr);
+    code_expressiont code{qex.where()};
     DATA_INVARIANT(
-      !has_subexpr(expr, ID_side_effect),
-      "the front-end should check quantified expressions for side-effects");
+      !has_subexpr(expr, ID_side_effect) ||
+        (code.operands()[0].id() == ID_side_effect &&
+         code.operands()[0].get_named_sub()[ID_statement].id() ==
+           ID_statement_expression),
+      "quantifier must not contain side effects");
+
+    // Handle the case that quantifier body is a statement expression.
+    if(
+      code.operands()[0].id() == ID_side_effect &&
+      code.operands()[0].get_named_sub()[ID_statement].id() ==
+        ID_statement_expression)
+    {
```

Comment:
Can you please move the code that's in this if branch to a method of its own? It has a very specific purpose and the overall code will be much easier to read when the `clean_expr` code can just invoke a named method.